vector/src/main/java/im/vector/activity/WidgetActivity.kt

```diff
@@ -20,355 +20,114 @@ package im.vector.activity
-                val cookieManager = android.webkit.CookieManager.getInstance()
-                cookieManager.setAcceptThirdPartyCookies(mWidgetWebView, true)
-            }
         }
-    }
 
-    private fun loadUrl() {
-        showWaitingView()
-        widgetsManager.getFormattedWidgetUrl(this, mWidget!!, object : ApiCallback<String> {
-            override fun onSuccess(url: String) {
-                mIsRefreshingToken = false
+        viewModel.selectSubscribe(this, RoomWidgetViewModelState::widgetName) { name ->
+            supportActionBar?.title = name
+        }
 
-                hideWaitingView()
-                mWidgetWebView.loadUrl(url)
-            }
+        viewModel.selectSubscribe(this, RoomWidgetViewModelState::canManageWidgets) {
+            invalidateOptionsMenu()
+        }
 
-            private fun onError(errorMessage: String) {
-                hideWaitingView()
-                toast(errorMessage)
-                finish()
+        viewModel.navigateEvent.observe(this, Observer { uxStateEvent ->
+            when (uxStateEvent?.getContentIfNotHandled()) {
+                RoomWidgetViewModel.NAVIGATE_FINISH -> {
+                    finish()
+                }
             }
+        })
 
-            override fun onNetworkError(e: Exception) {
-                onError(e.localizedMessage)
+        viewModel.toastMessageEvent.observe(this, Observer {
+            it?.getContentIfNotHandled()?.let {
+                Toast.makeText(this, it, Toast.LENGTH_LONG).show()
             }
+        })
+    }
 
-            override fun onMatrixError(e: MatrixError) {
-                onError(e.localizedMessage)
-            }
+    /* ==========================================================================================
+     * UI EVENTS
+     * ========================================================================================== */
 
-            override fun onUnexpectedError(e: Exception) {
-                onError(e.localizedMessage)
-            }
-        })
+    override fun onBackPressed() {
+        //Hugly i know
+        (supportFragmentManager.findFragmentByTag(FRAGMENT_TAG_WEBVIEW) as? RoomWidgetFragment)?.let { widgetFragment ->
+            if (widgetFragment.onBackPressed()) return
+        }
+        super.onBackPressed()
     }
 
+
     /* ==========================================================================================
      * companion
      * ========================================================================================== */
 
     companion object {
-        private val LOG_TAG = WidgetActivity::class.java.simpleName
 
+        private const val FRAGMENT_TAG_PERMISSION = "FRAGMENT_TAG_PERMISSION"
+        private const val FRAGMENT_TAG_WEBVIEW = "WebView"
         /**
          * The linked widget
          */
-        private const val EXTRA_WIDGET_ID = "EXTRA_WIDGET_ID"
+        const val EXTRA_WIDGET_ID = "EXTRA_WIDGET_ID"
```

Comment:
This should be private

Reply:
cannot be, it's used by the fragment also